plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java

```diff
@@ -735,7 +743,7 @@ public void rollback()
     public ConnectorInsertTableHandle beginInsert(ConnectorSession session, ConnectorTableHandle tableHandle, List<ColumnHandle> columns, RetryMode retryMode)
     {
         verify(!((JdbcTableHandle) tableHandle).isSynthetic(), "Not a table reference: %s", tableHandle);
-        if (retryMode != NO_RETRIES) {
+        if (retryMode != NO_RETRIES && !jdbcClient.supportsFaultTolerantExecution(session)) {
```

Comment:
Maybe move this check inside client itself. Now you are checking here, and there doing same check again inside to decide which code flow to follow - it is confusing.
Also please emit different error message if connector does not support retries at all and when connetor supports retries but `NON_TRANSACTIONAL_INSERT` is set. Currently we would get `This connector does not support query retries` for latter case and it is not a good error message.

Reply:
Also then maybe we do not need to expose `supportsFaultTolerantExecution()` from JdbcClient